src/compute/src/render/mod.rs

```diff
@@ -1120,6 +1142,40 @@ impl RenderTimestamp for Product<mz_repr::Timestamp, PointStamp<u64>> {
     }
 }
 
+/// A signal that can be awaited by operators to suspend them prior to startup.
+///
+/// Used to enforce a given configured hydration concurrency by only allowing that number of
+/// dataflows to perform hydration at the same time.
+///
+/// Note that currently only `persist_source` operators support suspension. Data can still enter a
+/// suspended dataflow through arrangement imports and ideally we would be able to suspend those as
+/// well.
+#[derive(Clone)]
+pub(super) struct StartSignal(
+    // The inner type is `Infallible` because no data is ever expected on this channel. Instead the
+    // signal is activated by dropping the corresponding `Sender`.
+    futures::future::Shared<oneshot::Receiver<Infallible>>,
```

Comment:
You could also use a broadcast channel: https://docs.rs/tokio/latest/tokio/sync/broadcast/fn.channel.html if you don't want to think about `Shared`'s semantics

Reply:
I had a `watch::Receiver` here first (I think that works the same as a broadcast channel for our purposes?) and couldn't get the `Future` impl to work.

I had something roughly like this:

```rust
#[derive(Clone)]
struct StartSignal(watch::Receiver<()>);

impl Future for StartSignal {
    type Output = ();

    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
        let wait_for_hangup = self.0.wait_for(|_| false).map(|_| ());
        pin!(wait_for_hangup).poll(cx)
    }
}
```

Somehow that didn't work as the task didn't get woken up when the corresponding `watch::Sender` was dropped. I figured it's probably not a good idea to create an ad-hoc future inside `poll` and I should instead create a future once and then store that. But I need the future to be clone-able, so I moved to `.shared()` and that removed the need for a `watch`.

Happy to learn about a better way to do this! My only requirement is having a `StartSignal` type I can `await`, so the various callsites don't have to be concerned with the implementation details of that signal.

Reply:
@teskje I think a broadcast channel would work here, `watch`'s have weird semantics! But for now `Shared` seems fine!